Hadrian uses wrong ar to build stage0 static archives

While cross-compiling, Hadrian uses the target's `ar` to build stage0 static archives. These archives target the host and therefore should be generated using the host's ar. This causes this failure,
```
/---------------------------------------------------------------------\
| Run Ar (stage = Stage0, package = ghc)                              |
|      input: _build/stage0/compiler/c/parser/cutils.o (and 464 more) |
|  => output: _build/stage0/compiler/libHSghc-8.3.a                   |
\---------------------------------------------------------------------/
/usr/bin/arm-linux-gnueabihf-ar q _build/stage0/compiler/libHSghc-8.3.a @/tmp/extra-file-81321174293242-1804289383846930886
/usr/bin/arm-linux-gnueabihf-ar: creating _build/stage0/compiler/libHSghc-8.3.a
/---------------------------------------------------\
| Successfully built library 'ghc' (Stage0, way v). |
| Library: _build/stage0/compiler/libHSghc-8.3.a    |
| Library synopsis: The GHC API.                    |
\---------------------------------------------------/
/-------------------------------------------------------------------------\
| Run Ghc LinkHs Stage0 (stage = Stage0, package = check-api-annotations) |
|      input: _build/stage0/utils/check-api-annotations/Main.o            |
|  => output: inplace/bin/check-api-annotations                           |
\-------------------------------------------------------------------------/
/mnt/work/ghc/roots/8.0.2/bin/ghc -Wall -hisuf hi -osuf o -hcsuf hc -static -hide-all-packages -no-user-package-db '-package-db /mnt/work/ghc/ghc/_build/stage0/bootstrapping.conf' '-package-id base-4.9.1.0' '-package-id containers-0.5.7.1' '-package-id Cabal-2.0.0.0' '-package-id directory-1.3.0.0' '-package-id ghc-8.3' -i -i_build/stage0/utils/check-api-annotations -i_build/stage0/utils/check-api-annotations/build/check-api-annotations/autogen -iutils/check-api-annotations/. -Iincludes -I_build/generated -I_build/stage0/utils/check-api-annotations -I/mnt/work/ghc/ghc/compiler -I/mnt/work/ghc/ghc/compiler/parser -I/mnt/work/ghc/ghc/compiler/utils -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/process-1.4.3.0/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/directory-1.3.0.0/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/unix-2.7.2.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/time-1.6.0.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/bytestring-0.10.8.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/base-4.9.1.0/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/integer-gmp-1.0.0.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/include -I_build/generated -optc-I_build/generated -optP-include -optP_build/stage0/utils/check-api-annotations/build/check-api-annotations/autogen/cabal_macros.h -optc-fno-stack-protector -odir _build/stage0/utils/check-api-annotations -hidir _build/stage0/utils/check-api-annotations -stubdir _build/stage0/utils/check-api-annotations -no-auto-link-packages -rtsopts _build/stage0/utils/check-api-annotations/Main.o -o inplace/bin/check-api-annotations -O -H32m -optc-Werror=unused-but-set-variable -optc-Wno-error=inline -Wall -XHaskell2010
/mnt/work/ghc/ghc/_build/stage0/compiler/libHSghc-8.3.a: error adding symbols: Archive has no index; run ranlib to add one
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
shakeArgsWith         0.000s    0%                           
Function shake        0.316s   15%  =====                    
Database read         0.355s   16%  ======                   
Database compression  0.079s    3%  =                        
With database         0.013s    0%                           
Running rules         1.330s   63%  =========================
Total                 2.094s  100%                           
Error when running Shake build system:
* inplace/bin/check-api-annotations
user error (Development.Shake.cmd, system command failed
Command: /mnt/work/ghc/roots/8.0.2/bin/ghc -Wall -hisuf hi -osuf o -hcsuf hc -static -hide-all-packages -no-user-package-db '-package-db /mnt/work/ghc/ghc/_build/stage0/bootstrapping.conf' '-package-id base-4.9.1.0' '-package-id containers-0.5.7.1' '-package-id Cabal-2.0.0.0' '-package-id directory-1.3.0.0' '-package-id ghc-8.3' -i -i_build/stage0/utils/check-api-annotations -i_build/stage0/utils/check-api-annotations/build/check-api-annotations/autogen -iutils/check-api-annotations/. -Iincludes -I_build/generated -I_build/stage0/utils/check-api-annotations -I/mnt/work/ghc/ghc/compiler -I/mnt/work/ghc/ghc/compiler/parser -I/mnt/work/ghc/ghc/compiler/utils -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/process-1.4.3.0/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/directory-1.3.0.0/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/unix-2.7.2.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/time-1.6.0.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/bytestring-0.10.8.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/base-4.9.1.0/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/integer-gmp-1.0.0.1/include -I/mnt/work/ghc/roots/8.0.2/lib/ghc-8.0.2/include -I_build/generated -optc-I_build/generated -optP-include -optP_build/stage0/utils/check-api-annotations/build/check-api-annotations/autogen/cabal_macros.h -optc-fno-stack-protector -odir _build/stage0/utils/check-api-annotations -hidir _build/stage0/utils/check-api-annotations -stubdir _build/stage0/utils/check-api-annotations -no-auto-link-packages -rtsopts _build/stage0/utils/check-api-annotations/Main.o -o inplace/bin/check-api-annotations -O -H32m -optc-Werror=unused-but-set-variable -optc-Wno-error=inline -Wall -XHaskell2010
Exit code: 1
Stderr:
/mnt/work/ghc/ghc/_build/stage0/compiler/libHSghc-8.3.a: error adding symbols: Archive has no index; run ranlib to add one
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
)

```

--- a/cfg/system.config.in
+++ b/cfg/system.config.in
@@ -16,6 +16,7 @@ make           = @MakeCmd@
 nm             = @NmCmd@
 objdump        = @ObjdumpCmd@
 ranlib         = @REAL_RANLIB_CMD@
+system-ar      = @AR_STAGE0@
 system-cc      = @CC_STAGE0@
 system-ghc     = @WithGhc@
 system-ghc-pkg = @GhcPkgCmd@

--- a/src/Builder.hs
+++ b/src/Builder.hs
@@ -28,7 +28,7 @@ data GhcPkgMode = Init | Update deriving (Eq, Generic, Show)
 -- @GhcPkg Stage0@ is the bootstrapping @GhcPkg@.
 -- @GhcPkg Stage1@ is the one built in Stage0.
 data Builder = Alex
-             | Ar
+             | Ar Stage
              | DeriveConstants
              | Cc CcMode Stage
              | Configure FilePath

--- a/src/Oracles/Path.hs
+++ b/src/Oracles/Path.hs
@@ -24,7 +24,8 @@ getTopDirectory = lift topDirectory
 systemBuilderPath :: Builder -> Action FilePath
 systemBuilderPath builder = case builder of
     Alex            -> fromKey "alex"
-    Ar              -> fromKey "ar"
+    Ar Stage0       -> fromKey "system-ar"
+    Ar _            -> fromKey "ar"
     Cc  _  Stage0   -> fromKey "system-cc"
     Cc  _  _        -> fromKey "cc"
     -- We can't ask configure for the path to configure!

--- a/src/Rules/Gmp.hs
+++ b/src/Rules/Gmp.hs
@@ -25,7 +25,7 @@ gmpMakefile = gmpBuildPath -/- "Makefile"
 
 configureEnvironment :: Action [CmdOption]
 configureEnvironment = sequence [ builderEnvironment "CC" $ Cc CompileC Stage1
-                                , builderEnvironment "AR" Ar
+                                , builderEnvironment "AR" (Ar Stage1)
                                 , builderEnvironment "NM" Nm ]
 
 gmpRules :: Rules ()
@@ -43,7 +43,7 @@ gmpRules = do
             putBuild "| No GMP library/framework detected; in tree GMP will be built"
             need [gmpLibrary]
             createDirectory gmpObjects
-            build $ Target gmpContext Ar [gmpLibrary] [gmpObjects]
+            build $ Target gmpContext (Ar Stage1) [gmpLibrary] [gmpObjects]
             copyFile (gmpBuildPath -/- "gmp.h") header
             copyFile (gmpBuildPath -/- "gmp.h") gmpLibraryInTreeH
 

--- a/src/Rules/Libffi.hs
+++ b/src/Rules/Libffi.hs
@@ -31,7 +31,7 @@ configureEnvironment = do
     sequence [ builderEnvironment "CC" $ Cc CompileC Stage1
              , builderEnvironment "CXX" $ Cc CompileC Stage1
              , builderEnvironment "LD" Ld
-             , builderEnvironment "AR" Ar
+             , builderEnvironment "AR" (Ar Stage1)
              , builderEnvironment "NM" Nm
              , builderEnvironment "RANLIB" Ranlib
              , return . AddEnv  "CFLAGS" $ unwords  cFlags ++ " -w"

--- a/src/Rules/Library.hs
+++ b/src/Rules/Library.hs
@@ -72,8 +72,8 @@ buildPackageLibrary context@Context {..} = do
 
         asuf <- libsuf way
         let isLib0 = ("//*-0" ++ asuf) ?== a
-        if isLib0 then build $ Target context Ar []   [a] -- TODO: Scan for dlls
-                  else build $ Target context Ar objs [a]
+        if isLib0 then build $ Target context (Ar stage) []   [a] -- TODO: Scan for dlls
+                  else build $ Target context (Ar stage) objs [a]
 
         synopsis <- interpretInContext context $ getPkgData Synopsis
         unless isLib0 . putSuccess $ renderLibrary

--- a/src/Settings/Builders/GhcCabal.hs
+++ b/src/Settings/Builders/GhcCabal.hs
@@ -25,7 +25,7 @@ ghcCabalBuilderArgs = builder GhcCabal ? do
             , packageConstraints
             , withStaged $ Cc CompileC
             , notStage0 ? with Ld
-            , with Ar
+            , withStaged Ar
             , with Alex
             , with Happy
             , verbosity < Chatty ? append [ "-v0", "--configure-option=--quiet"
@@ -91,7 +91,7 @@ cppArgs = arg $ "-I" ++ generatedPath
 
 withBuilderKey :: Builder -> String
 withBuilderKey b = case b of
-    Ar         -> "--with-ar="
+    Ar _       -> "--with-ar="
     Ld         -> "--with-ld="
     Cc  _ _    -> "--with-gcc="
     Ghc _ _    -> "--with-ghc="

--- a/src/Util.hs
+++ b/src/Util.hs
@@ -53,7 +53,7 @@ customBuild rs opts target@Target {..} = do
     withResources rs $ do
         putInfo target
         quietlyUnlessVerbose $ case builder of
-            Ar -> do
+            Ar _ -> do
                 output <- interpret target getOutput
                 if "//*.a" ?== output
                 then arCmd path argList

